### DISCUSSION TITLE
FORCE Level 1 CSD Image Quality ?

### DISCUSSION THREAD

**Discussion by Ekeany** [davidfrantz/force#94] [Q&A] [answered]
Hi Guys,

I am having some difficulty reproducing a quality higher level dataset for a specific  tile **(29UPV)**. 

In December I used the following query to generate 60 or so Sentinel-2 images:

`force-level1-sentinel2 /home/ubuntu/newvolume/level1/ /home/ubuntu/newvolume/level1/Dublin.txt "-6.2893295/53.3650992,-6.2903595/53.3230811,-6.2199783/53.3222609,-6.2192917/53.3644846,-6.2893295/53.3650992" 2020-01-01 2020-12-31 0 70`

With these images I was able to produce a higher level force dataset and everything worked perfectly !.

------------------------------------------------------------------------------------------------------------------------------------------

I am now in the process of recreating this dataset, however none of the tiles I want are online anymore (I guess I got lucky previously). So I have switched over to using the FORCE Level 1 CSD and the google cloud platform.

`force-level1-csd -c 0,70 -d 20200101,20201231 -s S2A, S2B /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1/Dublin.txt -8.203157114995863/51.85325785693969,-8.203157114995863/51.803257856939695,-8.153157114995862/51.803257856939695,-8.153157114995862/51.85325785693969,-8.203157114995863/51.85325785693969`

This code works and retrieves the desired tiles. However when I use the exact same processing parameter files as before the results are not the same. I expect there to be 9 data cubed tiles X0000_Y0000 -> X0003_Y0003 as was previous although using the force-level1-csd I only got 2 tiles the X0000_Y0000, X0000_Y0001. 

At first I thought I may have misspecified something but the all the parameter files and DEM etc are identical !. So I started looking at the `force-level1-csd` downloaded tiles and noticed that the majority of them were slithers of the actual tile examples down below. 

So is the answer that there just isn't any good data for this tile over this time period using `force-level1-csd` ? What is the difference between the quality of the images as I was able to produce a really nice data set before for this time period !

Do I need to expand the data range for maybe 2 years ? Or have I done something completely wrong 

Any help would be much appreciated

![image](https://user-images.githubusercontent.com/50741830/118362439-713d7e80-b587-11eb-8da5-fbde302618f5.png)

**Comment by ernstste** [comment]
Hi Ekeany,

Are you sure that you have set your AOI correctly? I only get results from the tile 29UNT when using your parameters. 
If 29UPV is the S2 tile you are interested in, you can simply define this tile as AOI:
`force-level1-csd -c 0,70 -d 20200101,20201231 -s S2A, S2B /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1/Dublin.txt T29UPV`

This request results in 100 S2 scenes that you can download. The tile you're interested in is covered by three S2 orbits, 23, 123, and 80 (West to East). What you can see in the images you posted is the coverage by orbit 23, which only has a small overlap with the tile (see screenshot, tile 29UPV and orbit 23 highlighted). So it seems like you may only have downloaded scenes from that orbit so far. There should be 35, 28, and 37 scenes for orbits 23, 123, and 80, respectively.

![grafik](https://user-images.githubusercontent.com/15325433/118374985-0dcf4300-b5bf-11eb-8c82-a4d58e556871.png)

**Comment by Ekeany** [comment]
Hi Stefan,

Thanks for the quick response !.

Apologies I copy and pasted the wrong AOI into the discussion .

I did as you suggested and downloaded the 100 Sentinel 2 scenes with out any problem except for a Python 2 depreciation warning. I can't use the tile specific feature as you suggested as I get some weird lockfile error so I just used the AOI again.

`force-level1-csd  -c 0,70 -d 20200101,20201231 -s S2A,S2B /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1 /home/ubuntu/newvolume/level1/Dublin.txt "-6.2893295/53.3650992,-6.2903595/53.3230811,-6.2199783/53.3222609,-6.2192917/53.3644846,-6.2893295/53.3650992"
`

![image](https://user-images.githubusercontent.com/50741830/118394049-4829f600-b63a-11eb-9d9c-4fb70827567b.png)

-----------------------------------------------------------------------------------------------------------------------------------------------------

The problem still persists when I process the data to level 2 ARD, I expect there to be 9 data cubed tiles X0000_Y0000 -> X0003_Y0003 but instead I got only 2 tiles X0000_Y0000, X0000_Y0001 which seems to be the just the top left corner of the scene. I am using the exact same DEM and [Parameter file](https://github.com/Ekeany/Ireland_building_heights/blob/main/Sentinel-2/PreProcessing/force-level2-param.prm) when I created the same level 2 ARD data from the force-level-sentinel2 downloader which is why I am so confused.

`force-level1-sentinel2 /home/ubuntu/newvolume/level1/ /home/ubuntu/newvolume/level1/Dublin.txt "-6.2893295/53.3650992,-6.2903595/53.3230811,-6.2199783/53.3222609,-6.2192917/53.3644846,-6.2893295/53.3650992" 2020-01-01 2020-12-31 0 70`

From my understanding the data on the google cloud platform should be identical to that on the ESA Hub ?. I know Ireland doesn't have the best weather but I find it hard to believe there are no cloud free days !! especially when I got such a nice output using the force-level1-sentinel2 command.

Thanks,
Eoghan

**Reply by ernstste** [reply]
Hi Eoghan,

If you downloaded all the data, you will see that a lot of the scenes are from orbits 123 and 80, which cover parts outside of the area you describe. 

Two points to check:
1. Is your file queue missing scenes?
2. If not, why are scenes not processed? Check the log files to see what happened during level 2 processing.

Stefan

**Comment by davidfrantz** [comment]
Hi Eoghan,

to my best knowledge, Google mirrors all data from ESA. We didn't see any indications of the opposite yet.

As Stefan suggested, I also would assume that either the query (AOI) is still incorrect or the Level 2 processing failed due to some reason.

What does ``cat  /home/ubuntu/newvolume/log/*.log`` bring up?

Since the current develop version, you can also use ``force-level2-report`` to generate a handy HTML report.

Btw: which FORCE version are you running? If I remember correctly, we fixed the issue with locked files in force-level1-csd at some point.

Best,
David

**Comment by Ekeany** [comment]
Hi Guys,

Again thanks for the responses.
I am attaching the log file as I am not really sure how to interpret it !.

For my setup I am currently using a t2.large instance 8GB RAM 2 vCPUs on the AWS platform,  with ubuntu 18.04.5 LTS and FORCE 3.6.3 I did notice this warning in my nohup.output :

`parallel: SIGTERM received. No new jobs will be started.
parallel: Waiting for these 1 jobs to finish. Send SIGTERM again to stop now.
`

Should I scale up to a larger instance as in t2.xlarge 16GB RAM 4 vCPUs ?

-------------------------------------------------------------------------------------------------------------------------------------------

As you suggest I believe it must be something to do with the Level2 processing.

100 scenes for the T29UPV tile were successfully downloaded, these 100 files are also present in the Queue.txt and in the log folder.
Of these 100 files  they are broken down into those orbits as Stefan described 37:R080, 35:R023, 28:R123

When looking through the log files:

- 19 files were successful all of which are the R023 orbit, which is why I am only seeing to tiles in my output X0000_Y0000 & X0000_Y0001.
- 49 files were skipped 18:R080, 16:R023, 15:R123
- 32 files have just the file name in the log file no explanation as to what happened to them ?

Should I reinstall a newer version of FORCE its just strange as I used this exact same machine PARAM files DEM etc.. to process the force-level1-sentinel2  files I am not sure what is going wrong !

[logs.txt](https://github.com/davidfrantz/force/files/6494424/logs.txt)

**Comment by davidfrantz** [comment, marked as answer]
Hi Eoghan,

thanks for the information. This immediately clears things up :)

`parallel: SIGTERM received. No new jobs will be started. parallel: Waiting for these 1 jobs to finish. Send SIGTERM again to stop now.`

Some processes were killed by the kernel - most likely due to issues with insufficient RAM.

The processing of a single Sentinel-2 image might need up to 8GB of RAM, probably a tad more. Thus your AWS instance is simply too small.

The only images that succeeded were the ones from orbit R023, which only covers a tiny part of the tile. FORCE also needs less RAM for these images.

Your setting in the Level 2 parameter file makes this worse (please search for *Bad setup* on https://force-eo.readthedocs.io/en/latest/howto/l2-ard.html#tut-ard)

```
NPROC = 2
NTHREAD = 1
```

This starts 2 simultaneous processes, which both may consume 8GB RAM. Mayhem ;)
Try this one instead. If it still fails, you will need to increase the RAM of your VM.

```
NPROC = 1
NTHREAD = 2
```

Hope this helps,
David

**Reply by Ekeany** [reply]
Okay Perfect thanks David will give this a try and let you know how it goes.

If I spin up a t2.xlarge 16GB RAM 4 vCPUs what would be an ideal setting ?

```
NPROC = 1
NTHREAD = 2
```

or 

`NPROC = 2
NTHREAD = 2`

**Reply by davidfrantz** [reply]
If `NPROC = 1, NTHREAD = 2` works on 2vCPU / 8GB, then `NPROC = 2, NTHREAD = 2` should work on 4vCPU / 16GB.

If not, consider using `NPROC = 1, NTHREAD = 4` on 4vCPU / 16GB.

Cheers,
David

PS: if it works, please make sure to mark the thread as answered :)

**Reply by Ekeany** [reply]
Thanks for solving the issue guys re ran the level2 processing scripts and got the output I expected !. 
For anyone else out there who plans to use AWS the t2.xlarge instance (4vCPU / 16GB) with `NPROC = 1, NTHREAD = 4` was what worked for me.
